src/compositor.zig

```diff
@@ -0,0 +1,840 @@
+            pixel.Pixel{ .rgba = .{ .r = 13, .g = 228, .b = 223, .a = 229 } },
+            runPixel(bg_mul.asPixel(), pixel.RGB.fromPixel(fg.asPixel()).asPixel(), .in),
+        );
+        try testing.expectEqualDeep(
+            pixel.Pixel{ .rgba = .{ .r = 9, .g = 170, .b = 167, .a = 171 } },
+            runPixel(bg_mul.asPixel(), fg.multiply().asPixel(), .in),
+        );
+        try testing.expectEqualDeep(
+            pixel.Pixel{ .rgba = .{ .r = 9, .g = 170, .b = 167, .a = 171 } },
+            runPixel(bg_mul.asPixel(), pixel.Alpha8.fromPixel(fg.asPixel()).asPixel(), .in),
+        );
+        try testing.expectEqualDeep(
+            pixel.Pixel{ .rgba = .{ .r = 9, .g = 167, .b = 163, .a = 167 } },
+            runPixel(bg_mul.asPixel(), pixel.Alpha4.fromPixel(fg.asPixel()).asPixel(), .in),
+        );
+        try testing.expectEqualDeep(
+            pixel.Pixel{ .rgba = .{ .r = 8, .g = 152, .b = 148, .a = 152 } },
+            runPixel(bg_mul.asPixel(), pixel.Alpha2.fromPixel(fg.asPixel()).asPixel(), .in),
+        );
+        try testing.expectEqualDeep(
+            pixel.Pixel{ .rgba = .{ .r = 13, .g = 228, .b = 223, .a = 229 } },
+            runPixel(bg_mul.asPixel(), .{ .alpha1 = .{ .a = 1 } }, .in),
+        );
+    }
+
+    {
+        // Alpha8
+        const bg_alpha8 = pixel.Alpha8.fromPixel(bg.asPixel());
+        try testing.expectEqualDeep(
+            pixel.Pixel{ .alpha8 = .{ .a = 229 } },
+            runPixel(bg_alpha8.asPixel(), pixel.RGB.fromPixel(fg.asPixel()).asPixel(), .in),
+        );
+        try testing.expectEqualDeep(
+            pixel.Pixel{ .alpha8 = .{ .a = 171 } },
+            runPixel(bg_alpha8.asPixel(), fg.asPixel(), .in),
+        );
+        try testing.expectEqualDeep(
+            pixel.Pixel{ .alpha8 = .{ .a = 171 } },
+            runPixel(bg_alpha8.asPixel(), pixel.Alpha8.fromPixel(fg.asPixel()).asPixel(), .in),
+        );
+        try testing.expectEqualDeep(
+            pixel.Pixel{ .alpha8 = .{ .a = 167 } },
+            runPixel(bg_alpha8.asPixel(), pixel.Alpha4.fromPixel(fg.asPixel()).asPixel(), .in),
+        );
+        try testing.expectEqualDeep(
+            pixel.Pixel{ .alpha8 = .{ .a = 152 } },
+            runPixel(bg_alpha8.asPixel(), pixel.Alpha2.fromPixel(fg.asPixel()).asPixel(), .in),
+        );
+        try testing.expectEqualDeep(
+            pixel.Pixel{ .alpha8 = .{ .a = 229 } },
+            runPixel(bg_alpha8.asPixel(), .{ .alpha1 = .{ .a = 1 } }, .in),
+        );
+    }
+
+    {
+        // Alpha4
+        const bg_alpha4 = pixel.Alpha4.fromPixel(bg.asPixel());
+        try testing.expectEqualDeep(
+            pixel.Pixel{ .alpha4 = .{ .a = 14 } },
+            runPixel(bg_alpha4.asPixel(), pixel.RGB.fromPixel(fg.asPixel()).asPixel(), .in),
+        );
+        try testing.expectEqualDeep(
+            pixel.Pixel{ .alpha4 = .{ .a = 11 } },
+            runPixel(bg_alpha4.asPixel(), fg.asPixel(), .in),
+        );
+        try testing.expectEqualDeep(
```

Comment:
NOTE: some of these tests were changed from 10 to 11, etc, in Alpha4, so we have some rounding changes. I think this is fine though if we commit to doing composition always in 8 bpc. This did not break any tests too so likely it's not a big deal at all.